### PR TITLE
mpdecimal: support conan v2

### PR DESCRIPTION
**Pull request by hellozee**
Specify library name and version:  **mpdecimal/2.5.x**

The recipe required system autconf for a single invocation of autoreconf. I could reason out that so I removed that line, and the recipe seems to work without any problems. It could be remnants of past but on the other hand it could be for a specific corner case which I am not aware of.

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
You need to update the imports for the bot to merge 

https://github.com/conan-io/conan-center-index/pull/16487/files#diff-b9e52b3ab97bd09c4f8f52f893b4c364e839013b474dc20d0484c5b3d973117eR1

**Comment by hellozee** [comment]
I honestly have no idea of what I am doing with this

**Review by sorny92 — COMMENTED** [review]
I've tested locally as it is in Windows and Linux with success. At this point I have no idea why it's failing.

At least now it's conan2? 🤷🏼

File: recipes/mpdecimal/2.5.x/conandata.yml [hunk 1]
```diff
@@ -7,14 +7,14 @@ sources:
     url: "http://www.bytereef.org/software/mpdecimal/releases/mpdecimal-2.5.0.tar.gz"
 patches:
   "2.5.1":
-    - base_path: "source_subfolder"
+    - base_path: "./"
```
Comment (prince-chrismc) [inline recipes/mpdecimal/2.5.x/conandata.yml:10]: This is not needed with the new layouts 

```suggestion
```

File: recipes/mpdecimal/2.5.x/conanfile.py [hunk 2]
```diff
@@ -25,123 +32,128 @@ class MpdecimalConan(ConanFile):
...
-            self.build_requires("automake/1.16.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-                self.build_requires("msys2/cci.latest")
+    def layout(self):
+        basic_layout(self, src_folder="src")
 
     def validate(self):
-        if self._is_msvc and self.settings.arch not in ("x86", "x86_64"):
+        if is_msvc(self) and self.settings.arch not in ("x86", "x86_64"):
             raise ConanInvalidConfiguration("Arch is unsupported")
```
Comment (prince-chrismc) [inline recipes/mpdecimal/2.5.x/conanfile.py:59]:
```suggestion
            raise ConanInvalidConfiguration(f"{self.ref} currently does not supported {self.settings.arch}. Contributions are welcomed")
```

**Comment by sorny92** [comment]
@prince-chrismc When you come back to review, have a look specially at [`5794649` (#16487)](https://github.com/conan-io/conan-center-index/pull/16487/commits/5794649353fae609ba243b9cf14e5f671cbb6acf) as it doesn't make sense to me but it's the only way I found to make it work locally for conan 1 and 2.

File: recipes/mpdecimal/2.5.x/test_v1_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
+conan_basic_setup(TARGETS)
+
+# Workaround for -rpath on arm64: https://cmake.org/cmake/help/latest/release/3.20.html#id2
```
Comment (prince-chrismc) [inline recipes/mpdecimal/2.5.x/test_v1_package/CMakeLists.txt:7]: There's a fix apple name hepler in conan for this AFAIK 🤔

File: recipes/mpdecimal/2.5.x/test_package/conanfile.py [hunk 4]
```diff
@@ -1,19 +1,48 @@
...
         cmake.build()
 
     def test(self):
-        if not tools.cross_building(self):
-            self.run("{} 13 100".format(os.path.join("bin", "test_package")), run_environment=True)
-            if self.options["mpdecimal"].cxx:
-                self.run("{} 13 100".format(os.path.join("bin", "test_package_cpp")), run_environment=True)
+        if can_run(self):
+            if self.settings.os != "Windows":
+                bin_path = pathlib.Path(self.cpp.build.bindirs[0], "bin", "test_package")
```
Comment (prince-chrismc) [inline recipes/mpdecimal/2.5.x/test_package/conanfile.py:39]: This looks like a bad copy where the package has too many `bin` folders 🤔

Comment (sorny92) [reply recipes/mpdecimal/2.5.x/test_package/conanfile.py]: It's a weird mess, if I remove the bin folder Linux build fails, if I add it the Windows one fails but nothing in the code told me why as the executable was where it was expected.

File: recipes/mpdecimal/2.5.x/test_v1_package/CMakeLists.txt (hunk 3 above)
Comment (sorny92) [reply recipes/mpdecimal/2.5.x/test_v1_package/CMakeLists.txt]: Looks like this is fixed now?

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
I think this is supper good a for first pass and we can do tweeks in the next PR :)

**Comment by hellozee** [comment]
> Well done @hellozee

\* Takes a look at the commit history * You might want to say that to a mirror, 😉

**Comment by prince-chrismc** [comment]
Thanks both of you, this was a tough one but I am sure lots of people will be very appreciative :)

**Review by SSE4 — APPROVED** [review]

**Pull request merged**